d3a_api_client/redis_client_base.py

```diff
@@ -2,57 +2,73 @@
 import logging
 import uuid
 from concurrent.futures.thread import ThreadPoolExecutor
-from functools import wraps
 
 from d3a_interface.utils import wait_until_timeout_blocking, key_in_dict_and_not_none
 from redis import StrictRedis
+from slugify import slugify
 
 from d3a_api_client import APIClientInterface
 from d3a_api_client.constants import MAX_WORKER_THREADS
+from d3a_api_client.utils import execute_function_util
 
 
 class RedisAPIException(Exception):
     pass
 
 
-def registered_connection(f):
-    @wraps(f)
-    def wrapped(self, *args, **kwargs):
-        if not self.is_active:
-            raise RedisAPIException(f'Registration has not completed yet.')
-        return f(self, *args, **kwargs)
-
-    return wrapped
-
-
-class RedisClient(APIClientInterface):
+class RedisClientBase(APIClientInterface):
     def __init__(self, area_id, autoregister=True, redis_url='redis://localhost:6379',
                  pubsub_thread=None):
         super().__init__(area_id, autoregister, redis_url)
+        self.area_uuid = None
+        self.area_slug = slugify(area_id, to_lower=True)
         self.redis_db = StrictRedis.from_url(redis_url)
         self.pubsub = self.redis_db.pubsub() if pubsub_thread is None else pubsub_thread
         self.area_id = area_id
         self.device_uuid = None
         self.is_active = False
         self._blocking_command_responses = {}
+        self._transaction_id_buffer = []
+        self._subscribed_aggregator_response_cb = None
         self._subscribe_to_response_channels(pubsub_thread)
         self.executor = ThreadPoolExecutor(max_workers=MAX_WORKER_THREADS)
         if autoregister:
             self.register(is_blocking=True)
 
+    @property
+    def _channel_prefix(self):
+        return f"{self.area_slug}"
+
     @property
     def channel_subs(self):
         return {
-            f"{self.area_id}/response/register_participant": self._on_register,
-            f"{self.area_id}/response/unregister_participant": self._on_unregister,
+            f'{self._channel_prefix}/response/register_participant': self._on_register,
+            f'{self._channel_prefix}/response/unregister_participant': self._on_unregister,
+            f'{self._channel_prefix}/*': self._on_event_or_response
         }
 
     def _subscribe_to_response_channels(self, pubsub_thread=None):
         self.pubsub.psubscribe(**self.channel_subs)
+        if b'aggregator_response' in self.pubsub.patterns:
+            self._subscribed_aggregator_response_cb = self.pubsub.patterns[b'aggregator_response']
```

Comment:
What kind of object is `self.pubsub.patterns[b'aggregator_response']`? Is it a callable?

Reply:
Sorry, this was messed up by me in the refactorization process. I will fix the whole `_subscribe_to_response_channels` actually.

Reply:
Since tests didn't fail because of this, it means that these functions are currently not tested, right? It would be great to do it so that we can spot errors automatically in the future.

Reply:
Good catch! This error was introduced because of a wrong implementation of `_check_transaction_id_cached_out` which should be
```
    def _check_transaction_id_cached_out(self, transaction_id):
        return transaction_id not in self._transaction_id_buffer
```
Nobody complained, because the registration process was alway happening. Even though the response was not handled correctly.
As we already test all this code in the integrationtests, I can not think of a way to test this specific case. 
I fixed the above mentioned function and also some logging in the redis client. Should be working now. 
let me know if you have an Idea how to test it. Thanks a lot!